README.md
# bip32_hdac

Forked [bip32 1.0.5](https://pub.dev/packages/bip32) compatible library for Flutter writing in Dart.

[bip32 1.0.5](https://pub.dev/packages/bip32) is different from github.com/cosmos/go-bip39 during the deriver process.

## Example
```dart

import 'package:bip32_hdac/bip32.dart' as bip32;
import 'package:hex/hex.dart';

main() {
  bip32.BIP32 node = bip32.BIP32.fromBase58('xprv9s21ZrQH143K3QTDL4LXw2F7HEK3wJUD2nW2nRk4stbPy6cq3jPPqjiChkVvvNKmPGJxWUtg6LnF5kejMRNNU3TGtRBeJgk33yuGBxrMPHi');

  print(HEX.encode(node.privateKey));
  // => e8f32e723decf4051aefac8e2c93c9c5b214313817cdb01a1494b917c8436b35

  bip32.BIP32 nodeNeutered = node.neutered();
  print(nodeNeutered.isNeutered());
  // => true

  print(HEX.encode(nodeNeutered.publicKey));
  // => 0339a36013301597daef41fbe593a02cc513d0b55527ec2df1050e2e8ff49c85c2

  print(nodeNeutered.toBase58());
  // => xpub661MyMwAqRbcFtXgS5sYJABqqG9YLmC4Q1Rdap9gSE8NqtwybGhePY2gZ29ESFjqJoCu1Rupje8YtGqsefD265TMg7usUDFdp6W1EGMcet8

  bip32.BIP32 child = node.derivePath('m/0/0');
  print(child.toBase58());
  // => xprv9ww7sMFLzJMzur2oEQDB642fbsMS4q6JRraMVTrM9bTWBq7NDS8ZpmsKVB4YF3mZecqax1fjnsPF19xnsJNfRp4RSyexacULXMKowSACTRc

  print(HEX.encode(child.privateKey));
  // => f26cf12f89ab91aeeb8d7324a22e8ba080829db15c9245414b073a8c342322aa

  bip32.BIP32 childNeutered = child.neutered();
  print(childNeutered.isNeutered());
  // => true

  print(HEX.encode(childNeutered.publicKey));
  // => 02756de182c5dd4b717ea87e693006da62dbb3cddaa4a5cad2ed1f5bbab755f0f5

  print(childNeutered.toBase58());
  // => xpub6AvUGrnEpfvJ8L7GLRkBTByQ9uBvUHp9o5VxHrFxhvzV4dSWkySpNaBoLR9FpbnwRmTa69yLHF3QfcaxbWT7gWdwws5k4dpmJvqpEuMWwnj

  bip32.BIP32 nodeFromSeed = bip32.BIP32.fromSeed(HEX.decode("000102030405060708090a0b0c0d0e0f"));
  print(nodeFromSeed.toBase58());
  // => xprv9s21ZrQH143K3QTDL4LXw2F7HEK3wJUD2nW2nRk4stbPy6cq3jPPqjiChkVvvNKmPGJxWUtg6LnF5kejMRNNU3TGtRBeJgk33yuGBxrMPHi

  bip32.BIP32 nodeFromPub = bip32.BIP32.fromBase58("xpub661MyMwAqRbcFtXgS5sYJABqqG9YLmC4Q1Rdap9gSE8NqtwybGhePY2gZ29ESFjqJoCu1Rupje8YtGqsefD265TMg7usUDFdp6W1EGMcet8");
  print(nodeFromPub.toBase58());
  // => xpub661MyMwAqRbcFtXgS5sYJABqqG9YLmC4Q1Rdap9gSE8NqtwybGhePY2gZ29ESFjqJoCu1Rupje8YtGqsefD265TMg7usUDFdp6W1EGMcet8

  var message = HEX.decode("0202020202020202020202020202020202020202020202020202020202020202");
  var signature = nodeFromSeed.sign(message);
  print(signature);
  // => [63, 219, 20, 114, 95, 184, 192, 55, 216, 206, 126, 121, 17, 71, 64, 70, 163, 82, 247, 73, 243, 95, 30, 137, 177, 155, 100, 225, 177, 203, 217, 147, 122, 64, 208, 129, 54, 133, 113, 41, 216, 160, 191, 15, 136, 98, 235, 25, 219, 178, 70, 222, 127, 151, 135, 242, 25, 192, 161, 187, 187, 84, 81, 215]

  print(HEX.encode(signature));
  // => 3fdb14725fb8c037d8ce7e7911474046a352f749f35f1e89b19b64e1b1cbd9937a40d08136857129d8a0bf0f8862eb19dbb246de7f9787f219c0a1bbbb5451d7

  print(nodeFromSeed.verify(message, signature));
  // => true
}

```


CHANGELOG.md
## 1.0.0-dev.1.1

- Initial version

test/bip32_test.dart
import 'dart:typed_data';
import '../lib/bip32_hdac.dart';
import 'package:hex/hex.dart';
import 'package:test/test.dart';
import 'dart:io';
import 'dart:convert';
final LITECOIN = new NetworkType(
  bip32: new Bip32Type(
    private: 0x019d9cfe,
    public: 0x019da462
  ),
  wif: 0xb0
);
List<dynamic> validAll = [];

void main() {
  Map<String, dynamic> fixtures = json.decode(File('./test/fixtures.json').readAsStringSync(encoding: utf8));
  (fixtures['valid'] as List<dynamic>).forEach((f) {
    f['master']['network'] = f['network'];
    f['master']['children'] = f['children'];
    f['master']['comment'] = f['comment'];
    (f['children'] as List<dynamic>).forEach((fc) {
      fc['network'] = f['network'];
      validAll.add(fc);
    });
    validAll.add(f['master']);
    validAll.forEach((ff) {
      group(ff['comment'] ?? ff['base58Priv'], () {
        setUp(() {});
        var network;
        if (ff['network'] == 'litecoin') {
         network = LITECOIN;
        }
        var hdPrv = BIP32.fromBase58(ff['base58Priv'], network);
        test('works for private key -> HD wallet', () {
          verify(hdPrv, true, ff, network);
        });

        var hdPub = BIP32.fromBase58(ff['base58'], network);
        test('works for public key -> HD wallet', () {
          verify(hdPub, false, ff, network);
        });

        if (ff['seed'] != null) {
          var seed = HEX.decode(ff['seed']);
          var hdSeed = BIP32.fromSeed(seed, network);
          test('works for seed -> HD wallet', () {
            verify(hdSeed, true, ff, network);
          });
        }
      });
    });
  });
  test('fromBase58 throws', () {
    (fixtures['invalid']['fromBase58'] as List<dynamic>).forEach((f) {
      var network;
      if (f['network'] != null && f['network'] == 'litecoin') network = LITECOIN;
      BIP32 hd;
      try {
        hd = BIP32.fromBase58(f['string'], network);
      } catch(err) {
        expect((err as ArgumentError).message, f['exception']);
      } finally {
        expect(hd, null);
      }
    });
  });
  test('works for Private -> public (neutered)', () {
    final f = fixtures['valid'][1];
    final c = f['master']['children'][0];
    final master = BIP32.fromBase58(f['master']['base58Priv'] as String);
    final child = master.derive(c['m']).neutered();
    expect(child.toBase58(), c['base58']);
  });
  test('works for Private -> public (neutered, hardened)', () {
    final f = fixtures['valid'][0];
    final c = f['master']['children'][0];
    final master = BIP32.fromBase58(f['master']['base58Priv'] as String);
    final child = master.deriveHardened(c['m']).neutered();
    expect(child.toBase58(), c['base58']);
  });
  test('works for Public -> public', () {
    final f = fixtures['valid'][1];
    final c = f['master']['children'][0];
    final master = BIP32.fromBase58(f['master']['base58'] as String);
    final child = master.derive(c['m']);
    expect(child.toBase58(), c['base58']);
  });
  test('throws on Public -> public (hardened)', () {
    final f = fixtures['valid'][0];
    final c = f['master']['children'][0];
    final master = BIP32.fromBase58(f['master']['base58'] as String);
    BIP32 hd;
    try {
      hd = master.deriveHardened(c['m']);
    } catch(err) {
      expect((err as ArgumentError).message, "Missing private key for hardened child key");
    } finally {
      expect(hd, null);
    }
  });
  test('throws on wrong types', () {
    final f = fixtures['valid'][0];
    final master = BIP32.fromBase58(f['master']['base58'] as String);
    (fixtures['invalid']['derive'] as List<dynamic>).forEach((fx) {
      var hd;
      try {
        hd = master.derive(fx);
      } catch (err) {
        expect((err as ArgumentError).message, "Expected UInt32");
      } finally {
        expect(hd, null);
      }
    });
    (fixtures['invalid']['deriveHardened'] as List<dynamic>).forEach((fx) {
      var hd;
      try {
        hd = master.deriveHardened(fx);
      } catch (err) {
        expect((err as ArgumentError).message, "Expected UInt31");
      } finally {
        expect(hd, null);
      }
    });
    (fixtures['invalid']['derivePath'] as List<dynamic>).forEach((fx) {
      var hd;
      try {
        hd = master.derivePath(fx);
      } catch (err) {
        expect((err as ArgumentError).message, "Expected BIP32 Path");
      } finally {
        expect(hd, null);
      }
    });
    var hdFPrv1, hdFPrv2;
    final ZERO32 = Uint8List.fromList(List.generate(32, (index) => 0));
    final ONE32 = Uint8List.fromList(List.generate(32, (index) => 1));
    try {
      hdFPrv1 = BIP32.fromPrivateKey(new Uint8List(2), ONE32);
    } catch (err) {
      expect((err as ArgumentError).message, "Expected property privateKey of type Buffer(Length: 32)");
    } finally {
      expect(hdFPrv1, null);
    }
    try {
      hdFPrv2 = BIP32.fromPrivateKey(ZERO32, ONE32);
    } catch (err) {
      expect((err as ArgumentError).message, "Private key not in range [1, n]");
    } finally {
      expect(hdFPrv2, null);
    }
  });
  test("works when private key has leading zeros", () {
    const key = "xprv9s21ZrQH143K3ckY9DgU79uMTJkQRLdbCCVDh81SnxTgPzLLGax6uHeBULTtaEtcAvKjXfT7ZWtHzKjTpujMkUd9dDb8msDeAfnJxrgAYhr";
    BIP32 hdkey = BIP32.fromBase58(key);
    expect(HEX.encode(hdkey.privateKey), "00000055378cf5fafb56c711c674143f9b0ee82ab0ba2924f19b64f5ae7cdbfd");
    BIP32 child = hdkey.derivePath("m/44'/0'/0'/0/0'");
    expect(HEX.encode(child.privateKey), "3348069561d2a0fb925e74bf198762acc47dce7db27372257d2d959a9e6f8aeb");
  });
  test("fromSeed", () {
    (fixtures['invalid']['fromSeed'] as List<dynamic>).forEach((f) {
      var hd;
      try {
        hd = BIP32.fromSeed(HEX.decode(f['seed']));
      } catch (err) {
        expect((err as ArgumentError).message, f['exception']);
      } finally {
        expect(hd, null);
      }
    });
  });
  test("ecdsa", () {
    Uint8List seed = Uint8List.fromList(List.generate(32, (index) => 1));
    Uint8List hash = Uint8List.fromList(List.generate(32, (index) => 2));
    String sigStr = "9636ee2fac31b795a308856b821ebe297dda7b28220fb46ea1fbbd7285977cc04c82b734956246a0f15a9698f03f546d8d96fe006c8e7bd2256ca7c8229e6f5c";
    Uint8List signature = HEX.decode(sigStr);
    BIP32 node = BIP32.fromSeed(seed);
    expect(HEX.encode(node.sign(hash)), sigStr);
    expect(node.verify(hash, signature), true);
    expect(node.verify(seed, signature), false);
  });

  test("derive", () {
    Uint8List seed = HEX.decode("5f13b32be2492c3a5806b11c5ece047047f3fc3f38159f291a9b6c4ff5dd504a8a50d46e72200406061955245d07bdd8dc6666916753fc7fd32a0195a2d78988");
    BIP32 root = BIP32.fromSeed(seed);
    root = root.derivePath("m/44'/1217'/0'/0/0");
    expect(HEX.encode(root.privateKey), "52351cb4356d9f374a9e25bfb57911b1d8b410d9196af174b5051e7cead0b8c6");

    Uint8List seed2 = HEX.decode("c2c51156714b94b21bc3ad48b865d3e86e2cdcb93f1990da5986bfbbffdec5e282ac825d9f2b86da566324a4d5bdd760315778820f8dc181e56d905549eec801");
    BIP32 root2 = BIP32.fromSeed(seed2);
    root2 = root2.derivePath("m/44'/1217'/0'/0/0");
    expect(HEX.encode(root2.privateKey), "56307d38deb07ecb11f5d38ced3eefa164e4fc2f6293119f0c765b3b3b9153b0");

    Uint8List seed3 = HEX.decode("f5db3b36cbbd80b9988cd01eda04333ab8159db2b53e79f34ca8ca09caeca47e3d666fb93a19bbd7d7fec13933d2a3fce2582b32ac63e1755b7f789db499a73f");
    BIP32 root3 = BIP32.fromSeed(seed3);
    root3 = root3.derivePath("m/44'/1217'/0'/0/0");
    expect(HEX.encode(root3.privateKey), "bd0e8f48c1bde66914073b5b20279481f27dce36ec2c164bca29944f3f6a08ad");
  });
}
void verify(BIP32 hd, prv, f, network) {
  expect(HEX.encode(hd.chainCode), f['chainCode']);
  expect(hd.depth, f['depth'] == null ? 0 : f['depth']);
  expect(hd.index, f['index'] == null ? 0 : f['index']);
  expect(HEX.encode(hd.fingerprint), f['fingerprint']);
  expect(HEX.encode(hd.identifier), f['identifier']);
  expect(HEX.encode(hd.publicKey), f['pubKey']);
  if (prv) {
    expect(hd.toBase58(), f['base58Priv']);
    expect(HEX.encode(hd.privateKey), f['privKey']);
    expect(hd.toWIF(), f['wif']);
  } else {
    expect(hd.privateKey, null);
  }
  expect(hd.neutered().toBase58(), f['base58']);
  expect(hd.isNeutered(), !prv);

  if (f['children'] == null) return;
  if (!prv &&
      (f['children'] as List<dynamic>).map((fc) => fc['hardened']).contains(
          true)) return;

  (f['children'] as List<dynamic>).forEach((cf) {
    var chd = hd.derivePath(cf['path']);
    verify(chd, prv, cf, network);
    var chdNoM = hd.derivePath((cf['path'] as String).substring(2)); // no m/
    verify(chdNoM, prv, cf, network);
  });

  // test deriving path from successive children
  var shd = hd;
  (f['children'] as List<dynamic>).forEach((cf) {
    if (cf['m'] == null) return;
    if (cf['hardened'] != null && cf['hardened'] as bool) {
      shd = shd.deriveHardened(cf['m']);
    } else {
      // verify any publicly derived children
      if (cf['base58'] != null) verify(
          shd.neutered().derive(cf['m']), false, cf, network);
      shd = shd.derive(cf['m']);
      verify(shd, prv, cf, network);
    }
  });
}



lib/src/bip32_base.dart
import 'dart:typed_data';
import 'utils/crypto.dart';
import 'utils/ecurve.dart' as ecc;
import 'package:bs58check/bs58check.dart' as bs58check;
import 'utils/wif.dart' as wif;
import 'dart:convert';

// TODO: Put public facing types in this file.
class Bip32Type {
  int public;
  int private;
  Bip32Type({this.public, this.private});

}
class NetworkType {
  int wif;
  Bip32Type bip32;
  NetworkType({this.wif, this.bip32});
}

final _BITCOIN = new NetworkType(
  wif: 0x80,
  bip32: new Bip32Type(
    public: 0x0488b21e,
    private: 0x0488ade4
  )
);
const HIGHEST_BIT = 0x80000000;
const UINT31_MAX = 2147483647; // 2^31 - 1
const UINT32_MAX = 4294967295; // 2^32 - 1
/// Checks if you are awesome. Spoiler: you are.
class BIP32 { 
  Uint8List _d;
  Uint8List _Q;
  Uint8List chainCode;
  int depth = 0;
  int index = 0;
  NetworkType network;
  int parentFingerprint = 0x00000000;
  BIP32(this._d, this._Q, this.chainCode, this.network);
  Uint8List get publicKey {
    if (_Q == null) _Q = ecc.pointFromScalar(_d, true);
    return _Q;
  }
  Uint8List get privateKey => _d;
  Uint8List get identifier => hash160(publicKey);
  Uint8List get fingerprint => identifier.sublist(0, 4);
  bool isNeutered() {
    return this._d == null;
  }
  BIP32 neutered() {
    final neutered = BIP32.fromPublicKey(this.publicKey, this.chainCode, this.network);
    neutered.depth = this.depth;
    neutered.index = this.index;
    neutered.parentFingerprint = this.parentFingerprint;
    return neutered;
  }
  String toBase58() {
    final version = (!isNeutered()) ? network.bip32.private : network.bip32.public;
    Uint8List buffer = new Uint8List(78);
    ByteData bytes = buffer.buffer.asByteData();
    bytes.setUint32(0, version);
    bytes.setUint8(4, depth);
    bytes.setUint32(5, parentFingerprint);
    bytes.setUint32(9, index);
    buffer.setRange(13, 45, chainCode);
    if (!isNeutered()) {
      bytes.setUint8(45, 0);
      buffer.setRange(46, 78, privateKey);
    } else {
      buffer.setRange(45, 78, publicKey);
    }
    return bs58check.encode(buffer);
  }
  String toWIF() {
    if (privateKey == null) {
      throw new ArgumentError("Missing private key");
    }
    return wif.encode(new wif.WIF(
        version: network.wif,
        privateKey:  privateKey,
        compressed: true
    ));
  }
  BIP32 derive(int index) {
    if (index > UINT32_MAX || index < 0) throw new ArgumentError("Expected UInt32");
    final isHardened = index >= HIGHEST_BIT;
    Uint8List data = new Uint8List(37);
    if (isHardened) {
      if (isNeutered()) {
        throw new ArgumentError("Missing private key for hardened child key");
      }
      data[0] = 0x00;
      data.setRange(1, 33, privateKey);
      data.buffer.asByteData().setUint32(33, index);
    } else {
      data.setRange(0, 33, publicKey);
      data.buffer.asByteData().setUint32(33, index);
    }
    final I = hmacSHA512(chainCode, data);
    final IL = I.sublist(0, 32);
    final IR = I.sublist(32);
    if (!ecc.isPrivate(IL)) {
      return derive(index + 1);
    }
    BIP32 hd;
    if (!isNeutered()) {
      final ki = ecc.privateAdd(privateKey, IL);
      if (ki == null) return derive(index + 1);
      hd = BIP32.fromPrivateKey(ki, IR, network);
    } else {
      final ki = ecc.pointAddScalar(publicKey, IL, true);
      if (ki == null) return derive(index + 1);
      hd = BIP32.fromPublicKey(ki, IR, network);
    }
    hd.depth = depth + 1;
    hd.index = index;
    hd.parentFingerprint = fingerprint.buffer.asByteData().getUint32(0);
    return hd;
  }
  BIP32 deriveHardened(int index) {
    if (index > UINT31_MAX || index < 0) throw new ArgumentError("Expected UInt31");
    return this.derive(index + HIGHEST_BIT);
  }
  BIP32 derivePath(String path) {
    final regex = new RegExp(r"^(m\/)?(\d+'?\/)*\d+'?$");
    if (!regex.hasMatch(path)) throw new ArgumentError("Expected BIP32 Path");
    List<String> splitPath = path.split("/");
    if (splitPath[0] == "m") {
      if (parentFingerprint != 0) throw new ArgumentError("Expected master, got child");
      splitPath = splitPath.sublist(1);
    }
    return splitPath.fold(this, (BIP32 prevHd,String indexStr) {
      int index;
      if (indexStr.substring(indexStr.length - 1) == "'") {
        index = int.parse(indexStr.substring(0, indexStr.length - 1));
        return prevHd.deriveHardened(index);
      } else {
        index = int.parse(indexStr);
        return prevHd.derive(index);
      }
    });
  }
  sign(Uint8List hash) {
    return ecc.sign(hash, privateKey);
  }
  verify(Uint8List hash, Uint8List signature) {
    return ecc.verify(hash, publicKey, signature);
  }
  factory BIP32.fromBase58(String string, [NetworkType nw]) {
    Uint8List buffer = bs58check.decode(string);
    if (buffer.length != 78) throw new ArgumentError("Invalid buffer length");
    NetworkType network = nw ?? _BITCOIN;
    ByteData bytes = buffer.buffer.asByteData();
    // 4 bytes: version bytes
    var version = bytes.getUint32(0);
    if (version != network.bip32.private && version != network.bip32.public) {
      throw new ArgumentError("Invalid network version");
    }
    // 1 byte: depth: 0x00 for master nodes, 0x01 for level-1 descendants, ...
    var depth = buffer[4];

    // 4 bytes: the fingerprint of the parent's key (0x00000000 if master key)
    var parentFingerprint = bytes.getUint32(5);
    if (depth == 0) {
      if (parentFingerprint != 0x00000000) throw new ArgumentError("Invalid parent fingerprint");
    }

    // 4 bytes: child number. This is the number i in xi = xpar/i, with xi the key being serialized.
    // This is encoded in MSB order. (0x00000000 if master key)
    var index = bytes.getUint32(9);
    if (depth == 0 && index != 0) throw new ArgumentError("Invalid index");

    // 32 bytes: the chain code
    Uint8List chainCode = buffer.sublist(13, 45);
    BIP32 hd;

    // 33 bytes: private key data (0x00 + k)
    if (version == network.bip32.private) {
      if (bytes.getUint8(45) != 0x00) throw new ArgumentError("Invalid private key");
      Uint8List k = buffer.sublist(46, 78);
      hd = BIP32.fromPrivateKey(k, chainCode, network);
    } else {
      // 33 bytes: public key data (0x02 + X or 0x03 + X)
      Uint8List X = buffer.sublist(45, 78);
      hd = BIP32.fromPublicKey(X, chainCode, network);
    }
    hd.depth = depth;
    hd.index = index;
    hd.parentFingerprint = parentFingerprint;
    return hd;
  }
  factory BIP32.fromPublicKey(Uint8List publicKey, Uint8List chainCode, [NetworkType nw]) {
    NetworkType network = nw ?? _BITCOIN;
    if (!ecc.isPoint(publicKey)) {
      throw new ArgumentError("Point is not on the curve");
    }
    return new BIP32(null, publicKey, chainCode, network);
  }
  factory BIP32.fromPrivateKey(Uint8List privateKey, Uint8List chainCode, [NetworkType nw]) {
    NetworkType network = nw ?? _BITCOIN;
    if (privateKey.length != 32) throw new ArgumentError("Expected property privateKey of type Buffer(Length: 32)");
    if (!ecc.isPrivate(privateKey)) throw new ArgumentError("Private key not in range [1, n]");
    return new BIP32(privateKey, null, chainCode, network);
  }
  factory BIP32.fromSeed(Uint8List seed, [NetworkType nw]) {
    if (seed.length < 16) {
      throw new ArgumentError("Seed should be at least 128 bits");
    }
    if (seed.length > 64) {
      throw new ArgumentError("Seed should be at most 512 bits");
    }
    NetworkType network = nw ?? _BITCOIN;
    final I = hmacSHA512(utf8.encode("Bitcoin seed"), seed);
    final IL = I.sublist(0, 32);
    final IR = I.sublist(32);
    return BIP32.fromPrivateKey(IL, IR, network);
  }
}




lib/src/utils/ecurve.dart
import 'dart:typed_data';
import 'package:hex/hex.dart';
import "package:pointycastle/ecc/curves/secp256k1.dart";
import "package:pointycastle/api.dart" show PrivateKeyParameter, PublicKeyParameter;
import 'package:pointycastle/ecc/api.dart' show ECPrivateKey, ECPublicKey, ECSignature, ECPoint;
import "package:pointycastle/signers/ecdsa_signer.dart";
import 'package:pointycastle/macs/hmac.dart';
import "package:pointycastle/digests/sha256.dart";
import 'package:pointycastle/src/utils.dart';
final ZERO32 = Uint8List.fromList(List.generate(32, (index) => 0));
final EC_GROUP_ORDER = HEX.decode("fffffffffffffffffffffffffffffffebaaedce6af48a03bbfd25e8cd0364141");
final EC_P = HEX.decode("fffffffffffffffffffffffffffffffffffffffffffffffffffffffefffffc2f");
final secp256k1 = new ECCurve_secp256k1();
final n = secp256k1.n;
final G = secp256k1.G;
BigInt nDiv2 = n >> 1;
const THROW_BAD_PRIVATE = 'Expected Private';
const THROW_BAD_POINT = 'Expected Point';
const THROW_BAD_TWEAK = 'Expected Tweak';
const THROW_BAD_HASH = 'Expected Hash';
const THROW_BAD_SIGNATURE = 'Expected Signature';

bool isPrivate (Uint8List x) {
  if (!isScalar(x)) return false;
  return _compare(x, ZERO32) > 0 && // > 0
      _compare(x, EC_GROUP_ORDER) < 0; // < G
}
bool isPoint(Uint8List p) {
  if (p.length < 33) {
    return false;
  }
  var t = p[0];
  var x = p.sublist(1, 33);

  if (_compare(x, ZERO32) == 0) {
    return false;
  }
  if (_compare(x, EC_P) == 1) {
    return false;
  }
  try {
    decodeFrom(p);
  } catch(err) {
    return false;
  }
  if ((t == 0x02 || t == 0x03) && p.length == 33) {
    return true;
  }
  var y = p.sublist(33);
  if (_compare(y, ZERO32) == 0) {
    return false;
  }
  if (_compare(y, EC_P) == 1) {
    return false;
  }
  if (t == 0x04 && p.length == 65) {
    return true;
  }
  return false;
}
bool isScalar (Uint8List x) {
  return x.length == 32;
}
bool isOrderScalar (x) {
  if (!isScalar(x)) return false;
  return _compare(x, EC_GROUP_ORDER) < 0; // < G
}
bool isSignature (Uint8List value) {
  Uint8List r = value.sublist(0, 32);
  Uint8List s = value.sublist(32, 64);
  return value.length == 64 &&
  _compare(r, EC_GROUP_ORDER) < 0 &&
  _compare(s, EC_GROUP_ORDER) < 0;
}
bool _isPointCompressed (Uint8List p) {
  return p[0] != 0x04;
}
bool assumeCompression(bool value, Uint8List pubkey) {
  if (value == null && pubkey != null) return _isPointCompressed(pubkey);
  if (value == null) return true;
  return value;
}
Uint8List pointFromScalar(Uint8List d, bool _compressed) {
  if (!isPrivate(d)) throw new ArgumentError(THROW_BAD_PRIVATE);
  BigInt dd = fromBuffer(d);
  ECPoint pp = G * dd;
  if (pp.isInfinity) return null;
  return getEncoded(pp, _compressed);
}
Uint8List pointAddScalar(Uint8List p,Uint8List tweak, bool _compressed) {
  if (!isPoint(p)) throw new ArgumentError(THROW_BAD_POINT);
  if (!isOrderScalar(tweak)) throw new ArgumentError(THROW_BAD_TWEAK);
  bool compressed = assumeCompression(_compressed, p);
  ECPoint pp = decodeFrom(p);
  if (_compare(tweak, ZERO32) == 0) return getEncoded(pp, compressed);
  BigInt tt = fromBuffer(tweak);
  ECPoint qq = G * tt;
  ECPoint uu = pp + qq;
  if (uu.isInfinity) return null;
  return getEncoded(uu, compressed);
}
Uint8List privateAdd (Uint8List d,Uint8List tweak) {
  if (!isPrivate(d)) throw new ArgumentError(THROW_BAD_PRIVATE);
  if (!isOrderScalar(tweak)) throw new ArgumentError(THROW_BAD_TWEAK);
  BigInt dd = fromBuffer(d);
  BigInt tt = fromBuffer(tweak);
  Uint8List dt = toBuffer((dd + tt) % n);
  Uint8List dt2 = new Uint8List(32);
  dt2.setRange(32 - dt.length, 32, dt);
  dt = dt2;
  
  if (!isPrivate(dt)) return null;
  return dt;
}
Uint8List sign(Uint8List hash, Uint8List x) {
  if (!isScalar(hash)) throw new ArgumentError(THROW_BAD_HASH);
  if (!isPrivate(x)) throw new ArgumentError(THROW_BAD_PRIVATE);
  ECSignature sig = deterministicGenerateK(hash, x);
  Uint8List buffer = new Uint8List(64);
  buffer.setRange(0, 32, encodeBigInt(sig.r));
  var s;
  if (sig.s.compareTo(nDiv2) > 0) {
    s = n - sig.s;
  } else {
    s = sig.s;
  }
  buffer.setRange(32, 64, encodeBigInt(s));
  return buffer;
}
bool verify(Uint8List hash, Uint8List q, Uint8List signature) {
  if (!isScalar(hash)) throw new ArgumentError(THROW_BAD_HASH);
  if (!isPoint(q)) throw new ArgumentError(THROW_BAD_POINT);
  // 1.4.1 Enforce r and s are both integers in the interval [1, n − 1] (1, isSignature enforces '< n - 1')
  if (!isSignature(signature)) throw new ArgumentError(THROW_BAD_SIGNATURE);

  ECPoint Q = decodeFrom(q);
  BigInt r = fromBuffer(signature.sublist(0, 32));
  BigInt s = fromBuffer(signature.sublist(32, 64));

  final signer = new ECDSASigner(null, new HMac(new SHA256Digest(), 64));
  signer.init(false, new PublicKeyParameter(new ECPublicKey(Q, secp256k1)));
  return signer.verifySignature(hash, new ECSignature(r, s));
  /* STEP BY STEP
  // 1.4.1 Enforce r and s are both integers in the interval [1, n − 1] (2, enforces '> 0')
  if (r.compareTo(n) >= 0) return false;
  if (s.compareTo(n) >= 0) return false;

  // 1.4.2 H = Hash(M), already done by the user
  // 1.4.3 e = H
  BigInt e = fromBuffer(hash);

  BigInt sInv = s.modInverse(n);
  BigInt u1 = (e * sInv) % n;
  BigInt u2 = (r * sInv) % n;

  // 1.4.5 Compute R = (xR, yR)
  //               R = u1G + u2Q
  ECPoint R = G * u1 + Q * u2;

  // 1.4.5 (cont.) Enforce R is not at infinity
  if (R.isInfinity) return false;

  // 1.4.6 Convert the field element R.x to an integer
  BigInt xR = R.x.toBigInteger();

  // 1.4.7 Set v = xR mod n
  BigInt v = xR % n;

  // 1.4.8 If v = r, output "valid", and if v != r, output "invalid"
  return v.compareTo(r) == 0;
  */
}

BigInt fromBuffer(Uint8List d) { return decodeBigInt(d); }
Uint8List toBuffer(BigInt d) { return encodeBigInt(d); }
ECPoint decodeFrom(Uint8List P) { return secp256k1.curve.decodePoint(P); }
Uint8List getEncoded(ECPoint P, compressed) { return P.getEncoded(compressed); }

ECSignature deterministicGenerateK(Uint8List hash, Uint8List x) {
  final signer = new ECDSASigner(null, new HMac(new SHA256Digest(), 64));
  var pkp = new PrivateKeyParameter(new ECPrivateKey(decodeBigInt(x), secp256k1));
  signer.init(true, pkp);
//  signer.init(false, new PublicKeyParameter(new ECPublicKey(secp256k1.curve.decodePoint(x), secp256k1)));
  return signer.generateSignature(hash);
}



int _compare(Uint8List a, Uint8List b) {
  BigInt aa = fromBuffer(a);
  BigInt bb = fromBuffer(b);
  if (aa == bb) return 0;
  if (aa > bb) return 1;
  return -1;
}


example/bip32_example.dart
import '../lib/bip32_hdac.dart' as bip32;
import 'package:hex/hex.dart';

main() {
  bip32.BIP32 node = bip32.BIP32.fromBase58('xprv9s21ZrQH143K3QTDL4LXw2F7HEK3wJUD2nW2nRk4stbPy6cq3jPPqjiChkVvvNKmPGJxWUtg6LnF5kejMRNNU3TGtRBeJgk33yuGBxrMPHi');

  print(HEX.encode(node.privateKey));
  // => e8f32e723decf4051aefac8e2c93c9c5b214313817cdb01a1494b917c8436b35

  bip32.BIP32 nodeNeutered = node.neutered();
  print(nodeNeutered.isNeutered());
  // => true

  print(HEX.encode(nodeNeutered.publicKey));
  // => 0339a36013301597daef41fbe593a02cc513d0b55527ec2df1050e2e8ff49c85c2

  print(nodeNeutered.toBase58());
  // => xpub661MyMwAqRbcFtXgS5sYJABqqG9YLmC4Q1Rdap9gSE8NqtwybGhePY2gZ29ESFjqJoCu1Rupje8YtGqsefD265TMg7usUDFdp6W1EGMcet8

  bip32.BIP32 child = node.derivePath('m/0/0');
  print(child.toBase58());
  // => xprv9ww7sMFLzJMzur2oEQDB642fbsMS4q6JRraMVTrM9bTWBq7NDS8ZpmsKVB4YF3mZecqax1fjnsPF19xnsJNfRp4RSyexacULXMKowSACTRc

  print(HEX.encode(child.privateKey));
  // => f26cf12f89ab91aeeb8d7324a22e8ba080829db15c9245414b073a8c342322aa

  bip32.BIP32 childNeutered = child.neutered();
  print(childNeutered.isNeutered());
  // => true

  print(HEX.encode(childNeutered.publicKey));
  // => 02756de182c5dd4b717ea87e693006da62dbb3cddaa4a5cad2ed1f5bbab755f0f5

  print(childNeutered.toBase58());
  // => xpub6AvUGrnEpfvJ8L7GLRkBTByQ9uBvUHp9o5VxHrFxhvzV4dSWkySpNaBoLR9FpbnwRmTa69yLHF3QfcaxbWT7gWdwws5k4dpmJvqpEuMWwnj

  bip32.BIP32 nodeFromSeed = bip32.BIP32.fromSeed(HEX.decode("000102030405060708090a0b0c0d0e0f"));
  print(nodeFromSeed.toBase58());
  // => xprv9s21ZrQH143K3QTDL4LXw2F7HEK3wJUD2nW2nRk4stbPy6cq3jPPqjiChkVvvNKmPGJxWUtg6LnF5kejMRNNU3TGtRBeJgk33yuGBxrMPHi

  bip32.BIP32 nodeFromPub = bip32.BIP32.fromBase58("xpub661MyMwAqRbcFtXgS5sYJABqqG9YLmC4Q1Rdap9gSE8NqtwybGhePY2gZ29ESFjqJoCu1Rupje8YtGqsefD265TMg7usUDFdp6W1EGMcet8");
  print(nodeFromPub.toBase58());
  // => xpub661MyMwAqRbcFtXgS5sYJABqqG9YLmC4Q1Rdap9gSE8NqtwybGhePY2gZ29ESFjqJoCu1Rupje8YtGqsefD265TMg7usUDFdp6W1EGMcet8

  var message = HEX.decode("0202020202020202020202020202020202020202020202020202020202020202");
  var signature = nodeFromSeed.sign(message);
  print(signature);
  // => [63, 219, 20, 114, 95, 184, 192, 55, 216, 206, 126, 121, 17, 71, 64, 70, 163, 82, 247, 73, 243, 95, 30, 137, 177, 155, 100, 225, 177, 203, 217, 147, 122, 64, 208, 129, 54, 133, 113, 41, 216, 160, 191, 15, 136, 98, 235, 25, 219, 178, 70, 222, 127, 151, 135, 242, 25, 192, 161, 187, 187, 84, 81, 215]

  print(HEX.encode(signature));
  // => 3fdb14725fb8c037d8ce7e7911474046a352f749f35f1e89b19b64e1b1cbd9937a40d08136857129d8a0bf0f8862eb19dbb246de7f9787f219c0a1bbbb5451d7

  print(nodeFromSeed.verify(message, signature));
  // => true
}
